### PR TITLE
Add CPU usage nano cores for windows nodes

### LINKED ISSUE #76740: kubelet does not report `usageNanoCores` for node on Windows
**What happened**:

`usageNanoCores` is not reported for `node` on Windows.
```
{ 
  "node": {
   "nodeName": "e2e-test-yjhong-windows-node-group-wv9b",
   "systemContainers": [
    {
     "name": "pods",
     "startTime": "2019-04-17T22:58:59Z",
     "cpu": {
      "time": null
     },
     "memory": {
      "time": null
     },
     "userDefinedMetrics": null
    }
   ],
   "startTime": "2019-04-17T22:24:57Z",
   "cpu": {
    "time": "2019-04-17T22:58:58Z",
    "usageCoreNanoSeconds": 55040000000
   },
   "memory": {
    "time": "2019-04-17T22:58:58Z",
    "availableBytes": 7619317760,
    "usageBytes": 1213886464,
    "workingSetBytes": 433328128,
    "rssBytes": 0,
    "pageFaults": 0,
    "majorPageFaults": 0
   },
...
```

**Environment**:
- Kubernetes version (use `kubectl version`): Kubernetes v1.15.0-alpha.1.76+d48509bdf01ee1-
- Cloud provider or hardware configuration: GCE
- OS (e.g: `cat /etc/os-release`): Windows 1809
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/cc @dashpole @feiskyer

### PR DESCRIPTION
**Pull request by liyanhui1228**
**What type of PR is this?**

> /kind feature

**Which issue(s) this PR fixes**:

Fixes #76740 

```release-note
cpuUsageNanoCores is now reported in the Kubelet summary API on Windows nodes
```

### CODE REVIEW

File: pkg/kubelet/winstats/perfcounters.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ const (
 	// Perf counters are updated every second. This is the same as the default cadvisor collection period
 	// see https://github.com/google/cadvisor/blob/master/docs/runtime_options.md#housekeeping
 	perfCounterUpdatePeriod = 1 * time.Second
+	// defaultCachePeriod is the default cache period for each cpuUsage.
+	defaultCachePeriod = 15 * time.Second
```
Comment (feiskyer) [inline pkg/kubelet/winstats/perfcounters.go]: Would 15s too long compared to other metrics? Note that `perfCounterUpdatePeriod` is 1 seconds.

File: pkg/kubelet/winstats/perfcounter_nodestats.go [hunk 2]
```diff
@@ -110,6 +116,23 @@ func (p *perfCounterNodeStatsClient) startMonitoring() error {
...
 
+	// Cache the CPU usage every 15 seconds
+	go wait.Forever(func() {
+		cpuCores := runtime.NumCPU()
+		cpuValue, err := cpuCounter.getData()
+		if err != nil {
+			klog.Errorf("Unable to get cpu perf counter data; err: %v", err)
+			return
+		}
+		newValue := p.nodeMetrics.cpuUsageCoreNanoSeconds + uint64((float64(cpuValue)/100.0)*float64(cpuCores)*1e9)
```
[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Could we also avoid converting back and forth between float64 and uint64 if we multiplied by 10 million at the end instead of 1e9 (e.g. combine converting from nano and dividing percent by 100)?

[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Can you also add a unit test for the math (maybe factor just the math out to a helper to make that easier).

**Review by mtaufen — COMMENTED** [review]
Thanks for fixing this! Had a few comments.

File: pkg/kubelet/winstats/perfcounters.go (hunk 1 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounters.go]: cAdvisor uses 15s as the time interval, and we want to be consistent with cAdvisor in here.

File: pkg/kubelet/winstats/perfcounter_nodestats.go [hunk 3]
```diff
@@ -110,6 +116,23 @@ func (p *perfCounterNodeStatsClient) startMonitoring() error {
 		p.collectMetricsData(cpuCounter, memWorkingSetCounter, memCommittedBytesCounter, networkAdapterCounter)
 	}, perfCounterUpdatePeriod)
 
+	// Cache the CPU usage every 15 seconds
+	go wait.Forever(func() {
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Yeah that's right, the update frequency is different in here.

And for the cpuUsageNanoCores it could have at most 14 seconds lag of the real value, as the goroutine runs every 15 seconds and if we query at the 29th second it will still return the value at the 15th second.

File: pkg/kubelet/winstats/perfcounters.go (hunk 1 above)
Comment (yujuhong) [reply pkg/kubelet/winstats/perfcounters.go]: @dashpole do you now where this is set in cadvisor?

Comment (dashpole) [reply pkg/kubelet/winstats/perfcounters.go]: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cadvisor/cadvisor_linux.go#L63

it is between 10 and 15 seconds

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: This calculation is actually the same as the [convertCPUValue](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/winstats/perfcounter_nodestats.go#L182), I changed to call it directly, and probably leave the calculation as it is for now?

And I just added the unit test for the calculation.

File: pkg/kubelet/winstats/perfcounter_nodestats.go [hunk 4]
```diff
@@ -110,6 +116,23 @@ func (p *perfCounterNodeStatsClient) startMonitoring() error {
 		p.collectMetricsData(cpuCounter, memWorkingSetCounter, memCommittedBytesCounter, networkAdapterCounter)
 	}, perfCounterUpdatePeriod)
 
+	// Cache the CPU usage every 15 seconds
```
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Done.

File: pkg/kubelet/winstats/perfcounters.go (hunk 1 above)
Comment (yujuhong) [reply pkg/kubelet/winstats/perfcounters.go]: Let's do 10s and add a comment here saying that this matches the cadvisor setting. 

10s also matches the period we use for containers.

File: pkg/kubelet/winstats/perfcounter_nodestats.go [hunk 5]
```diff
@@ -110,6 +116,22 @@ func (p *perfCounterNodeStatsClient) startMonitoring() error {
 		p.collectMetricsData(cpuCounter, memWorkingSetCounter, memCommittedBytesCounter, networkAdapterCounter)
 	}, perfCounterUpdatePeriod)
 
+	// Cache the CPU usage every defaultCachePeriod
+	go wait.Forever(func() {
+		cpuValue, err := cpuCounter.getData()
```
Comment (yujuhong) [inline pkg/kubelet/winstats/perfcounter_nodestats.go]: @feiskyer do you know how much time/resource it takes to get the data from the counter?

If it's bad, an alternative would be to cache a copy of `nodeMetrics` and compute the difference. WDYT?

File: pkg/kubelet/winstats/perfcounters.go (hunk 1 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounters.go]: Done.

**Comment by neolit123** [comment]
cc @kubernetes/sig-windows-misc

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 5 above)
Comment (feiskyer) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: not sure, @PatrickLang have you noticed the time for this?

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Ah, I see it's supposed to be an accumulator across the lifetime of the node. Ok then.

[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Thanks for adding a test.

File: pkg/kubelet/winstats/winstats_test.go [hunk 6]
```diff
@@ -123,6 +130,23 @@ func TestWinVersionInfo(t *testing.T) {
...
 }
 
+func convertCPUValue(cpuValue uint64) uint64 {
+	cpuCores := 4
+	f := fakeWinNodeStatsClient{}
+	nodeMetrics, err := f.getNodeMetrics()
+	if err != nil {
+		return 0
+	}
+	newValue := nodeMetrics.cpuUsageCoreNanoSeconds + uint64((float64(cpuValue)/100.0)*float64(cpuCores)*1e9)
```
Comment (mtaufen) [inline pkg/kubelet/winstats/winstats_test.go]: Rather than copy the calculation here, the test should call the same function used by the application code so they can't diverge.

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dashpole) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: @mtaufen I think you are correct here.  `cpuUsageCoreNanoSeconds ` is the cumulative usage over time.  UsageNanoCores, which is an average over the collection period, should be uncs(t2) - uncs(t1) / (t2-t1).  It looks like we are not dividing by the time between collections.

File: pkg/kubelet/winstats/winstats_test.go (hunk 6 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/winstats_test.go]: Done.

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Thanks for the explanation! I was misunderstanding the meaning of `UsageNanoCores` and didn't use the average value. I modified the calculation, please take another look.

[the remark this answers is not in the document]
Comment (dashpole) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: It looks correct to me now.

**Comment by dashpole** [comment]
This lgtm, thanks for the fix.

One interesting result of this design is that the average value is the only part not collected on-demand.  This means successive scrapes of the summary API could have different cumulative metrics, which when diffed produce a different rate than the rate given by the average metric.  The end result is that the two metrics (cumulative vs rate) would produce slightly different graphs for users.  The only remedies would be to either fetch all node CPU metrics from the cache (which would make that metric stale), or use the cumulative value collected at query time (which would use inconsistent periods) to calculate the rate.

Despite this, I think we should stick with the current design, as the above is only noticeable when comparing the two metrics.

/lgtm

**Comment by yujuhong** [comment]
Looks good. Could you squash the commits?

**Comment by liyanhui1228** [comment]
> Looks good. Could you squash the commits?

Done.

**Comment by yujuhong** [comment]
/approve

nit: a commit message with slightly more description about what the PR does would be nice :-)

File: pkg/kubelet/winstats/perfcounter_nodestats.go [hunk 7]
```diff
@@ -171,23 +189,29 @@ func (p *perfCounterNodeStatsClient) collectMetricsData(cpuCounter, memWorkingSe
...
 	// This converts perf counter data which is cpu percentage for all cores into nanoseconds.
 	// The formula is (cpuPercentage / 100.0) * #cores * 1e+9 (nano seconds). More info here:
 	// https://github.com/kubernetes/heapster/issues/650
 	newValue := p.nodeMetrics.cpuUsageCoreNanoSeconds + uint64((float64(cpuValue)/100.0)*float64(cpuCores)*1e9)
 	return newValue
 }
 
+func (p *perfCounterNodeStatsClient) getCPUUsageNanoCores() uint64 {
+	// UsageNanoCores is an average over the collection period
+	cpuUsageNanoCores := (p.cpuUsageCache.latestValue - p.cpuUsageCache.previousValue) / uint64(defaultCachePeriod)
```
Comment (mtaufen) [inline pkg/kubelet/winstats/perfcounter_nodestats.go]: Since time in Go is denominated in nanoseconds (a 10 second cache period is `10*1000*1000*1000` under the hood), will this actually give a value in cores rather than nanocores? 

Assuming 10 second period:
`uncs(t1) = 1*1e9`
`uncs(t2) = 3*1e9`
`uncs(t2) - uncs(t1) / t2 - t1 = 2*1e9 / 10*1e9 = 0.2` but with integer division it's just 0.

File: pkg/kubelet/winstats/winstats.go [hunk 8]
```diff
@@ -69,6 +70,11 @@ type nodeInfo struct {
 	startTime time.Time
 }
 
+type cpuUsageCache struct {
```
Comment (mtaufen) [inline pkg/kubelet/winstats/winstats.go]: Maybe rename this to cpuUsageCoreNanoSecondsCache to make it clear what value is stored in the cache.

File: pkg/kubelet/winstats/winstats_test.go [hunk 9]
```diff
@@ -123,6 +130,21 @@ func TestWinVersionInfo(t *testing.T) {
...
+	p := perfCounterNodeStatsClient{}
+	newValue := p.convertCPUValue(cpuCores, cpuValue)
+	assert.Equal(t, newValue, uint64(400000000000))
+}
+
+func TestGetCPUUsageNanoCores(t *testing.T) {
+	p := perfCounterNodeStatsClient{}
+	p.cpuUsageCache = cpuUsageCache{200000000000, 500000000000}
+	cpuUsageNanoCores := p.getCPUUsageNanoCores()
+	assert.Equal(t, cpuUsageNanoCores, uint64(1229782918))
```
Comment (mtaufen) [inline pkg/kubelet/winstats/winstats_test.go]: Based on (500... - 200...) / 10..., shouldn't we expect a 3... for the result?

File: pkg/kubelet/winstats/winstats_test.go [hunk 10]
```diff
@@ -123,6 +130,21 @@ func TestWinVersionInfo(t *testing.T) {
 		KernelVersion: "v42"})
 }
 
+func TestConvertCPUValue(t *testing.T) {
+	var cpuValue uint64 = 10000
```
Comment (mtaufen) [inline pkg/kubelet/winstats/winstats_test.go]: cpuValue is a percent so it's probably more realistic for this to be between 0 and 100. 

For this and below, can you also test some edge cases like 0 and 100, and some in-between value?

**Comment by mtaufen** [comment]
/hold
I had a couple more questions from my latest pass. Feel free to remove the hold once they're addressed.

**Comment by mtaufen** [comment]
Also, to get rid of the do-not-merge/release-note-label-needed label, add a release note to the bottom of the PR description, like:
````
```release-note
cpuUsageNanoCores is now reported in the Kubelet summary API on Windows nodes
```
````

File: pkg/kubelet/winstats/winstats_test.go (hunk 10 above)
Comment (mtaufen) [reply pkg/kubelet/winstats/winstats_test.go]: Also recommend using subtests for each test case: https://blog.golang.org/subtests

File: pkg/kubelet/winstats/winstats_test.go (hunk 9 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/winstats_test.go]: I was taking the value returned from the expected value and thanks for noticing this! I fixed the test data for `cpuUsageCache` to use `uint64` type as well as fixed the calculation and the expected value is matching with real calculation now.

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 7 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Thanks for pointing out this! I wasn't quite familiar with Go data types and I just fixed the calculation to 
```
cpuUsageNanoCores := (p.cpuUsageCache.latestValue - p.cpuUsageCache.previousValue) * uint64(time.Second / time.Nanosecond) / uint64(defaultCachePeriod)
```
And the windows unit test passed with the expected value `uint64(300000000)` for `cpuUsageCache{latestValue: uint64(5000000000), previousValue: uint64(2000000000),}`.
```
PASS
ok      _/C_/Windows/system32/kubernetes/pkg/kubelet/winstats      0.079s
```

File: pkg/kubelet/winstats/winstats_test.go (hunk 10 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/winstats_test.go]: Done.

File: pkg/kubelet/winstats/winstats.go (hunk 8 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/winstats.go]: Done.

File: pkg/kubelet/winstats/winstats_test.go [hunk 11]
```diff
@@ -123,6 +130,34 @@ func TestWinVersionInfo(t *testing.T) {
...
+		newValue := p.convertCPUValue(cpuCores, tc.cpuValue)
+		assert.Equal(t, newValue, tc.expected)
+	}
+}
+
+func TestGetCPUUsageNanoCores(t *testing.T) {
+	p := perfCounterNodeStatsClient{}
+	p.cpuUsageCoreNanoSecondsCache = cpuUsageCoreNanoSecondsCache{
+		latestValue:   uint64(5000000000),
+		previousValue: uint64(2000000000),
```
Comment (mtaufen) [inline pkg/kubelet/winstats/winstats_test.go]: Can you also test the cases where just previousValue is 0, and where both values are 0?

File: pkg/kubelet/winstats/perfcounter_nodestats.go [hunk 12]
```diff
@@ -171,23 +189,28 @@ func (p *perfCounterNodeStatsClient) collectMetricsData(cpuCounter, memWorkingSe
...
+func (p *perfCounterNodeStatsClient) convertCPUValue(cpuCores int, cpuValue uint64) uint64 {
 	// This converts perf counter data which is cpu percentage for all cores into nanoseconds.
 	// The formula is (cpuPercentage / 100.0) * #cores * 1e+9 (nano seconds). More info here:
 	// https://github.com/kubernetes/heapster/issues/650
 	newValue := p.nodeMetrics.cpuUsageCoreNanoSeconds + uint64((float64(cpuValue)/100.0)*float64(cpuCores)*1e9)
 	return newValue
 }
 
+func (p *perfCounterNodeStatsClient) getCPUUsageNanoCores() uint64 {
+	cpuUsageNanoCores := (p.cpuUsageCoreNanoSecondsCache.latestValue - p.cpuUsageCoreNanoSecondsCache.previousValue) * uint64(time.Second/time.Nanosecond) / uint64(defaultCachePeriod)
```
Comment (mtaufen) [inline pkg/kubelet/winstats/perfcounter_nodestats.go]: I think this is mathematically correct but I'd recommend adjusting to something like this:
```
cachePeriodSeconds := uint64(defaultCachePeriod/time.Second)
cpuUsageNanoCores := (p.cpuUsageCoreNanoSecondsCache.latestValue - p.cpuUsageCoreNanoSecondsCache.previousValue)/cachePeriodSeconds)
```
- While multiplying by time.Seconds in the numerator will cancel the time.Seconds in the denominator, the max value of `uint64` is around `18e18`, so `Xe9*1e9` may be getting into overflow territory, so I think it might be safer to divide the nanoseconds out of the denominator (especially since defaultCachePeriod is a known, fixed constant).
- `time.Second/time.Nanosecond = 1*1000*1000*1000/1` so you can drop that division.

**Review by mtaufen — COMMENTED** [review]
Looking pretty good, had just a couple more comments.

File: pkg/kubelet/winstats/winstats_test.go (hunk 11 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/winstats_test.go]: Done.

File: pkg/kubelet/winstats/perfcounter_nodestats.go (hunk 12 above)
Comment (liyanhui1228) [reply pkg/kubelet/winstats/perfcounter_nodestats.go]: Changed, thanks for the review!

**Comment by yujuhong** [comment]
@liyanhui1228  could you cherrypick this PR to 1.15 and 1.14 as a bug fix? Thanks!
